File: iosApp/iosApp/Features/Diary/notes/NotesView.swift
//
//  NotesView.swift
//  iosApp
//
//  Created by velkonost on 27.09.2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import SwiftUI
import SharedSDK

struct NotesView: View {
    
    @Binding var state: NotesViewState
    
    private let showAd: Bool
    private let adPosition: Int
    
    private var isLoading: Bool
    private let createGoalClick: () -> Void
    private let createNoteClick: () -> Void
    
    private let items: [Note]
    private let itemClick: (Note) -> Void
    private let itemLikeClick: (Note) -> Void
    private let onBottomReach: () -> Void
    
    init(
        showAd: Bool,
        adPosition: Int, state: Binding<NotesViewState>,
        isLoading: Bool, items: [Note],
        createGoalClick: @escaping () -> Void, createNoteClick: @escaping () -> Void,
        itemClick: @escaping (Note) -> Void, 
        itemLikeClick: @escaping (Note) -> Void,
        onBottomReach: @escaping () -> Void
    ) {
        self.showAd = showAd
        self.adPosition = adPosition
        self._state = state
        self.isLoading = isLoading
        self.items = items
        
        self.createGoalClick = createGoalClick
        self.createNoteClick = createNoteClick
        
        self.itemClick = itemClick
        self.itemLikeClick = itemLikeClick
        
        self.onBottomReach = onBottomReach
    }
    
    var body: some View {
        ZStack {
            if isLoading && items.isEmpty {
                Loader()
            } else {
                if items.isEmpty {
                    PlaceholderView(text: SharedR.strings().placeholder_diary_notes.desc().localized())
                } else {
                    ScrollView(showsIndicators: false) {
                        LazyVStack(spacing: 0) {
                            ForEach(0..<items.count, id: \.self) { index in
                                NoteItem(
                                    item: items[index],
                                    onClick: itemClick,
                                    onLikeClick: itemLikeClick
                                )
                                .onAppear {
                                    checkPaginationThreshold(currentItemId: items[index].id)
                                }
                                
                                if index != 0 && index % adPosition == 0 {
                                    AdView(showAd: showAd)
                                        .padding(.vertical, 2)
                                }
                            }
                            
                            if items.count < adPosition {
                                AdView(showAd: showAd)
                                    .padding(.vertical, 2)
                            }
                        }
                        .padding(.init(top: .zero, leading: 20, bottom: 100, trailing: 20))
                        
                        Loader().opacity(state.isLoading ? 1 : 0)
                    }
                    .fadingEdge()
                }
                
                VStack(alignment: .trailing) {
                    Spacer()
                    AddNoteItem {
                        createGoalClick()
                    } createNoteClick: {
                        createNoteClick()
                    }
                }
                .frame(width: UIScreen.screenWidth)
                
            }
        }
        .frame(minWidth: 0, maxWidth: .infinity, minHeight: 0, maxHeight: .infinity)
        
    }
}

extension NotesView {
    func checkPaginationThreshold(currentItemId: Int32) {
        let data = items
        let thresholdIndex = data.index(data.endIndex, offsetBy: -5)
        
        if data.firstIndex(where: { $0.id == currentItemId })! >= thresholdIndex && !isLoading {
            onBottomReach()
        }
    }
}

File: androidApp/features/diary/src/main/kotlin/com/velkonost/getbetter/android/features/diary/notes/NotesView.kt
package com.velkonost.getbetter.android.features.diary.notes

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.velkonost.getbetter.core.compose.components.Loader
import com.velkonost.getbetter.core.compose.components.PlaceholderView
import com.velkonost.getbetter.core.compose.components.ad.AdView
import com.velkonost.getbetter.core.compose.components.notelist.AddNoteItem
import com.velkonost.getbetter.core.compose.components.notelist.NoteItem
import com.velkonost.getbetter.core.compose.extensions.OnBottomReached
import com.velkonost.getbetter.core.compose.extensions.fadingEdge
import com.velkonost.getbetter.shared.core.model.note.Note
import com.velkonost.getbetter.shared.resources.SharedR
import dev.icerock.moko.resources.compose.stringResource

@Composable
fun NotesView(
    modifier: Modifier = Modifier,
    loadMorePrefetch: Int,
    isLoading: Boolean,
    items: List<Note>,
    adPosition: Int,
    adSlotId: Int,
    createGoalClick: () -> Unit,
    createNoteClick: () -> Unit,
    itemClick: (Note) -> Unit,
    itemLikeClick: (Note) -> Unit,
    onBottomReach: () -> Unit
) {
    val listState = rememberLazyListState()

    Box(modifier = modifier.fillMaxSize()) {
        if (isLoading && items.isEmpty()) {
            Loader(modifier = Modifier.align(Alignment.Center))
        } else {
            if (items.isEmpty()) {
                PlaceholderView(text = stringResource(resource = SharedR.strings.placeholder_diary_notes))
            } else {
                LazyColumn(
                    modifier = modifier
                        .fillMaxSize()
                        .fadingEdge(),
                    state = listState,
                    contentPadding = PaddingValues(bottom = 140.dp)
                ) {
                    itemsIndexed(items, key = { _, it -> it.id }) { index, item ->
                        NoteItem(
                            item = item,
                            onClick = itemClick,
                            onLikeClick = itemLikeClick
                        )

                        if (index % adPosition == 0 && index != 0) {
                            AdView(slotId = adSlotId)
                        }
                    }

                    if (items.size < adPosition) {
                        item {
                            AdView(slotId = adSlotId)
                        }
                    }

                    if (isLoading) {
                        item {
                            Box(modifier = modifier.fillMaxSize()) {
                                Loader(modifier = modifier.align(Alignment.Center))
                            }
                        }
                    }
                }
            }

            AddNoteItem(
                createGoalClick = createGoalClick,
                createNoteClick = createNoteClick
            )
        }
    }

    listState.OnBottomReached(
        buffer = loadMorePrefetch,
        isLoading = isLoading
    ) {
        onBottomReach.invoke()
    }
}